sklearn/metrics/pairwise.py

```diff
@@ -367,23 +387,29 @@ def _euclidean_distances_upcast_fast(X, XX, Y, YY):
         xe = xs + (chunk_size if i < n_chunks_X - 1 else n_samples_X_rem)
 
         X_chunk = X[xs:xe].astype(np.float64)
-        if not XX:
+        if XX is None:
             XX_chunk = row_norms(X_chunk, squared=True)
         else:
-            XX_chunk = XX[xs:xe].astype(np.float64)
+            XX_chunk = XX[xs:xe]
 
         for j in range(n_chunks_Y):
             ys = j * chunk_size
             ye = ys + (chunk_size if j < n_chunks_Y - 1 else n_samples_Y_rem)
 
             Y_chunk = Y[ys:ye].astype(np.float64)
-            if not YY:
+            if YY is None:
                 YY_chunk = row_norms(Y_chunk, squared=True)
             else:
-                YY_chunk = YY[ys:ye].astype(np.float64)
+                YY_chunk = YY[ys:ye]
 
-            d = - 2 * safe_sparse_dot(X_chunk, Y_chunk.T, dense_output=True)
-            _add_norms(d, XX_chunk, YY_chunk)
+            if X is Y and j < i:
+                # when X is Y the distance matrix is symmetric so we only need
+                # to compute half of it.
+                d = distances[ys:ye, xs:xe].T
```

Comment:
That's not a lot, but the way you implemented it might recompute the row norm when it won't be used because we copy the chunk above the diagonal.
I don't have a strong opinion toward avoiding it. But I note that it could be easily avoided by adding a `break` statement to stay under the diagonal and copy the chunk above it.

Reply:
fixed it